rules/directory/filter.bzl

```diff
@@ -0,0 +1,65 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Rules to filter files from a directory."""
+
+load(":providers.bzl", "DirectoryInfo")
+load(":utils.bzl", "get_relative")
+
+def _directory_filter_impl(ctx):
+    directory = ctx.attr.directory[DirectoryInfo]
+    srcs = depset([
+        get_relative(directory, path, require_file = True)
+        for path in ctx.attr.srcs
+    ])
+    data = [
+        get_relative(directory, path, require_file = True)
+        for path in ctx.attr.data
+    ]
+
+    return DefaultInfo(
+        files = srcs,
+        runfiles = ctx.runfiles(files = data, transitive_files = srcs),
+    )
+
+directory_filter = rule(
```

Comment:
Is it possible to implement directory_fileter using the glob rule in the other PR?

Reply:
Strictly speaking, this rule isn't necessary - we could just tell a user to use a glob all the time (and the performance should be equivalent - globs are trivially cheap if we don't have a `*` in them). The main reason I didn't was just because if you had `*` in the filename it'd fail, and on principle it seemed weird to require a glob.

I don't have particularly strong feelings about this, so I'm happy to change this if you'd like me to. Let's discuss what to do with this in our meeting today.

Reply:
In the meantime, I've pulled this out of the CL.